File: macos-app/Sources/ColoringBook/Network/Protocol.swift
import Foundation

// MARK: Outgoing (Client → Server)

enum ClientMessage: Encodable {
    case strokeStart(StrokeStartPayload)
    case strokePoint(strokeId: String, point: StrokePoint)
    case strokeEnd(strokeId: String)
    case cursor(x: Double, y: Double)
    case setPage(WirePage?)
    case clearCanvas

    private enum CodingKeys: String, CodingKey {
        case type, stroke, strokeId, point, x, y, page
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .strokeStart(let p):
            try c.encode("stroke_start", forKey: .type)
            try c.encode(p, forKey: .stroke)
        case .strokePoint(let strokeId, let point):
            try c.encode("stroke_point", forKey: .type)
            try c.encode(strokeId, forKey: .strokeId)
            try c.encode(point, forKey: .point)
        case .strokeEnd(let strokeId):
            try c.encode("stroke_end", forKey: .type)
            try c.encode(strokeId, forKey: .strokeId)
        case .cursor(let x, let y):
            try c.encode("cursor", forKey: .type)
            try c.encode(x, forKey: .x)
            try c.encode(y, forKey: .y)
        case .setPage(let page):
            try c.encode("set_page", forKey: .type)
            try c.encode(page, forKey: .page)
        case .clearCanvas:
            try c.encode("clear_canvas", forKey: .type)
        }
    }
}

/// A page of line art (transparent PNG) shared across a room. The `imageBase64`
/// may be an empty string for blank-paper pages.
struct WirePage: Codable, Equatable {
    let pageId: String
    let displayName: String
    let mimeType: String
    let imageBase64: String
}

struct StrokeStartPayload: Codable {
    let id: String
    let userId: String
    let tool: Tool
    let color: WireColor
    let brushSize: Double
    let point: StrokePoint
}

// MARK: Incoming (Server → Client)

struct ServerEnvelope: Decodable { let type: String }

struct WireStroke: Decodable {
    let id: String
    let userId: String
    let tool: Tool
    let color: WireColor
    let brushSize: Double
    let points: [StrokePoint]
    let complete: Bool

    func toStroke() -> Stroke {
        Stroke(id: id, userId: userId, tool: tool, color: color,
               brushSize: brushSize, points: points, complete: complete)
    }
}

struct WirePeer: Decodable {
    let userId: String
    let name: String
    let color: String
}

struct RoomStateMessage: Decodable {
    let strokes: [WireStroke]
    let peers: [WirePeer]
    let you: You
    let page: WirePage?
    struct You: Decodable { let userId: String }
}

struct PageChangedMessage: Decodable {
    let userId: String
    let page: WirePage?
}

struct CanvasClearedMessage: Decodable {
    let userId: String
}

struct PeerJoinedMessage: Decodable { let peer: WirePeer }
struct PeerLeftMessage: Decodable { let userId: String }

struct WireStrokeStart: Decodable {
    let id: String
    let userId: String
    let tool: Tool
    let color: WireColor
    let brushSize: Double
    let point: StrokePoint
}

struct StrokeStartMessage: Decodable {
    let userId: String
    let stroke: WireStrokeStart
}

struct StrokePointMessage: Decodable {
    let userId: String
    let strokeId: String
    let point: StrokePoint
}

struct StrokeEndMessage: Decodable {
    let userId: String
    let strokeId: String
}

struct CursorMessage: Decodable {
    let userId: String
    let x: Double
    let y: Double
}

File: server/src/protocol.ts
export type Point = {
  x: number;
  y: number;
  pressure: number;
  t: number;
};

export type Color = {
  r: number;
  g: number;
  b: number;
  a: number;
};

export type Tool = "sketchpen" | "pencil" | "watercolor" | "crayon" | "pastel";

export type StrokeHeader = {
  id: string;
  userId: string;
  tool: Tool;
  color: Color;
  brushSize: number;
};

export type Stroke = StrokeHeader & {
  points: Point[];
  complete: boolean;
};

export type Peer = {
  userId: string;
  name: string;
  color: string;
  cursor?: { x: number; y: number };
};

/// A page is a coloring sheet with its own image + its own strokes.
/// Each room owns multiple pages; only one is "active" at any time.
export type Page = {
  pageId: string;
  displayName: string;
  mimeType: string;       // "image/png"
  imageBase64: string;    // empty string for blank paper
};

/// On-the-wire stroke that knows which page it belongs to (so peers can
/// route incoming stroke events to the correct page's stroke list).
export type WireStroke = StrokeHeader & {
  pageId: string;
  points: Point[];
  complete: boolean;
};

export type ClientMessage =
  | { type: "stroke_start"; stroke: StrokeHeader & { point: Point; pageId: string } }
  | { type: "stroke_point"; strokeId: string; point: Point }
  | { type: "stroke_end"; strokeId: string }
  | { type: "cursor"; x: number; y: number }
  | { type: "add_page"; page: Page }                  // creates new + makes active
  | { type: "select_page"; pageId: string }            // switch active page
  | { type: "delete_page"; pageId: string }            // remove page (and its strokes)
  | { type: "clear_canvas" };                          // clears active page's strokes

export type ServerMessage =
  | {
      type: "room_state";
      peers: Peer[];
      you: { userId: string };
      pages: Page[];
      activePageId: string | null;
      strokes: WireStroke[];
    }
  | { type: "peer_joined"; peer: Peer }
  | { type: "peer_left"; userId: string }
  | {
      type: "stroke_start";
      userId: string;
      stroke: StrokeHeader & { point: Point; pageId: string };
    }
  | { type: "stroke_point"; userId: string; strokeId: string; point: Point }
  | { type: "stroke_end"; userId: string; strokeId: string }
  | { type: "cursor"; userId: string; x: number; y: number }
  | { type: "page_added"; userId: string; page: Page }
  | { type: "page_selected"; userId: string; pageId: string }
  | { type: "page_deleted"; userId: string; pageId: string }
  | { type: "canvas_cleared"; userId: string; pageId: string };
